### PR TITLE
Threaded file-writer status tracker.

### PR DESCRIPTION
**Pull request by SkyToGround**
As the title says.

### CODE REVIEW

File: file_writer_control/InThreadStatusTracker.py [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+        if current_schema == ANSW_IDENTIFIER:
+            self.process_answer(deserialise_answer(message))
+        elif current_schema == STAT_IDENTIFIER:
+            self.process_status(deserialise_status(message))
+        elif current_schema == STOP_TIME_IDENTIFIER:
+            self.process_set_stop_time(deserialise_stop_time(message))
+        elif current_schema == START_IDENTIFIER:
+            self.process_start(deserialise_start(message))
+        elif current_schema == STOPPED_IDENTIFIER:
+            self.process_stopped(deserialise_stopped(message))
```
Comment (umithardal) [inline file_writer_control/InThreadStatusTracker.py:69]: While not critical, if-elif sequences can be shorthanded via using Python dictionaries; i.e., while case-switch structure does not exist in Python, it is very easy to create. For example [see here](https://uwpce-pythoncert.github.io/PythonCertDevel/modules/DictionaryAsSwitch.html) which is very relevant to this code.

File: file_writer_control/InThreadStatusTracker.py [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+        else:
+            self.process_unknown(message)
+        self.send_status_if_updated(update_time)
+
+    def process_unknown(self, message: bytes):
+        pass  # Do nothing for now
```
Comment (umithardal) [inline file_writer_control/InThreadStatusTracker.py:75]: Is this a placeholder?

File: file_writer_control/InThreadStatusTracker.py [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+        """
+        for worker in self.known_workers.values():
+            if worker.last_update >= limit_time:
+                self.queue.put(worker)
+        for job in self.known_jobs.values():
+            if job.last_update >= limit_time:
+                self.queue.put(job)
+        for command in self.known_commands.values():
+            if command.last_update >= limit_time:
+                self.queue.put(command)
```
Comment (umithardal) [inline file_writer_control/InThreadStatusTracker.py:91]: These three for loops can be made into one via `zip`. That is we may write `for (x, y, z) in zip(k, l, m):`. Nothing to worry about if you find explicit loops more readable.

File: file_writer_control/InThreadStatusTracker.py [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+        current_state = extract_worker_state_from_status(status_update)
+        self.known_workers[status_update.service_id].state = current_state
+        if current_state == WorkerState.WRITING:
+            job_id = json.loads(status_update.status_json)["job_id"]
+            self.check_for_job_presence(job_id)
+            self.known_jobs[job_id].state = JobState.WRITING
+            try:
+                self.known_jobs[job_id].service_id = status_update.service_id
+            except RuntimeError:
+                pass  # Expected error, do nothing
```
Comment (umithardal) [inline file_writer_control/InThreadStatusTracker.py:180]: I did not understand what is happening here.

File: file_writer_control/InThreadStatusTracker.py (hunk 2 above)
Comment (SkyToGround) [reply file_writer_control/InThreadStatusTracker.py:75]: Fair point. I have deleted this function.

File: file_writer_control/InThreadStatusTracker.py (hunk 1 above)
Comment (SkyToGround) [reply file_writer_control/InThreadStatusTracker.py:69]: Doing it this way requires the use of functions or lambdas which then in turn results in the same number of lines. Still, I have re-written it now so I might as well leave it as is.

File: file_writer_control/InThreadStatusTracker.py (hunk 3 above)
Comment (SkyToGround) [reply file_writer_control/InThreadStatusTracker.py:91]: I do not actually much care for `zip` so unless you insist, I would like to leave the code as is.

File: file_writer_control/InThreadStatusTracker.py (hunk 4 above)
Comment (SkyToGround) [reply file_writer_control/InThreadStatusTracker.py:180]: I have added some comments. Tell me if this aids in understanding what this statement does.

**Comment by SkyToGround** [comment]
Oops. Forgot to push the changes to the PR yesterday. I have done so now.

File: file_writer_control/InThreadStatusTracker.py (hunk 3 above)
Comment (umithardal) [reply file_writer_control/InThreadStatusTracker.py:90]: No problem at all.

File: file_writer_control/InThreadStatusTracker.py (hunk 1 above)
Comment (umithardal) [reply file_writer_control/InThreadStatusTracker.py]: My intention was not to reduce the number of lines but rather to have a more Pythonic way. The current code with dictionary is certainly more elegant! Great job!

**Review by umithardal — APPROVED** [review]

**Pull request merged**